state/src/state.rs

```diff
@@ -0,0 +1,658 @@
+            }
+        }
+        self.add_checkpoint(address);
+
+        if let Some(ref mut state_object_entry) = self.cache.borrow_mut().get_mut(address) {
+            match state_object_entry.state_object {
+                Some(ref mut state_object) => {
+                    state_object.set_storage(key, value);
+                    state_object_entry.status = ObjectStatus::Dirty;
+                }
+                None => panic!("state object always exist in cache."),
+            }
+        }
+        Ok(())
+    }
+
+    pub fn insert_cache(&self, address: &Address, state_object_entry: StateObjectEntry) {
+        let is_dirty = state_object_entry.is_dirty();
+        let old_entry = self
+            .cache
+            .borrow_mut()
+            .insert(*address, state_object_entry.clone_dirty());
+
+        if is_dirty {
+            if let Some(checkpoint) = self.checkpoints.borrow_mut().last_mut() {
+                checkpoint.entry(*address).or_insert(old_entry);
+            }
+        }
+    }
+
+    pub fn commit(&mut self) -> Result<(), Error> {
+        assert!(self.checkpoints.borrow().is_empty());
+
+        // firstly, update account storage tree
+        for (_address, entry) in self
+            .cache
+            .borrow_mut()
+            .iter_mut()
+            .filter(|&(_, ref a)| a.is_dirty())
+        {
+            if let Some(ref mut state_object) = entry.state_object {
+                state_object.commit_storage(&mut self.db)?;
+                state_object.commit_code(&mut self.db)?;
+            }
+        }
+
+        // secondly, update the whold state tree
+        let mut trie = PatriciaTrie::from(&mut self.db, RLPNodeCodec::default(), &self.root.0)?;
+        for (address, entry) in self
+            .cache
+            .borrow_mut()
+            .iter_mut()
+            .filter(|&(_, ref a)| a.is_dirty())
+        {
+            entry.status = ObjectStatus::Committed;
+            match entry.state_object {
+                Some(ref mut state_object) => {
+                    trie.insert(address, &rlp::encode(&state_object.account()))?;
+                }
+                None => {
+                    trie.remove(address)?;
+                }
+            }
+        }
+        self.root = From::from(&trie.root()?[..]);
+        Ok(())
+    }
+
+    pub fn checkpoint(&mut self) -> usize {
+        let mut checkpoints = self.checkpoints.borrow_mut();
+        let index = checkpoints.len();
+        checkpoints.push(HashMap::new());
+        index
+    }
+
+    fn add_checkpoint(&self, address: &Address) {
+        if let Some(ref mut checkpoint) = self.checkpoints.borrow_mut().last_mut() {
+            checkpoint.entry(*address).or_insert_with(|| {
+                self.cache
+                    .borrow()
+                    .get(address)
+                    .map(StateObjectEntry::clone_dirty)
+            });
+        }
+    }
+
+    // If the transaction if executed successfully
```

Comment:
The discard_checkpoint will merge the last checkpoint to the prev, this will be called when the message call successfully.